doc/crctk.3
.TH "crctk" "3" "Sat 1 Mar 2014" "" ""
.SH NAME
.PP
crctk \- CRC32 Hexstring Toolkit
.SH SYNPOSIS
.PP
\f[B]crctk\f[] [\f[I]OPTIONS\f[]]...
[\f[I]\-\-\f[]] [\f[I]FILE\f[]|\f[I]FILE\-LIST\f[]]
.SH DESCRIPTION
.PP
crctk is a program for checking the integrity of files using the CRC32
checksum algorithm and generating CRC32 checksum data.
The CRC32 checksums can be input and output in various ways.
.PP
Similar to tools like sha1sum(1), md5sum(1) and others, \f[I]crctk\f[]
is also used to create database files containing mappings from filepaths
to checksums, which can be used later on to verify a batch of files.
crctk, however, uses a binary format \-\- libtinycdb, an implementation
of Daniel J.
Bernstein\[aq]s CDB \-\- to address a problem with POSIX\-compliant
filepaths in textual form: Because filepaths may contain every character
except \[aq]\[aq] and \[aq]/\[aq], a binary format drastically
simplifies the parsing of filenames, while the textfiles from sha1sum et
al.
do not allow filenames containing \[aq]\[aq], for example.
CDB in turn is very fast, simple and stable.
.SS HEXSTRING TAGS
.PP
Because of their short length, it is common for CRC32 checksums to be
included as hexadecimal strings in the filename of a \f[I]FILE\f[].
Such tags are known to crctk as matching the regular expression
\f[B][[:punct:]]\\?[[:xdigit:]]\\{8\\}[[:punct:]]\\?\f[], that is a
hexadecimal string optionally surrounded by punctuation characters,
including brackets (example: [11FFAABB]).
crctk is also used to manage such tags: it can tag files, update tags
and use the CRC32 information from tags to validate a file.
.SH COMMAND LINE
.SS GENERIC OPTIONS
.IP \[bu] 2
\f[B]\-c|\-\-calc [\-n|\-\-numerical] [\f[I]\-\-\f[]] \f[I]FILE\f[] \f[]
.PD 0
.P
.PD
Compute the CRC32 checksum of \f[I]FILE\f[], print it to stdout and
exit.
If \f[I]\-n\f[] is present, the numerical value will be output instead
of the hexstring.
.IP \[bu] 2
\f[B]\-e|\-\-tag\-regex \f[I]EXPRESSION\f[] \f[]
.PD 0
.P
.PD
Replace the built\-in regex used to match tags (see section:
\f[B]HEXSTRING TAGS\f[]) with \f[I]EXPRESSION\f[].
.IP \[bu] 2
\f[B]\-h|\-\-help\f[]
.PD 0
.P
.PD
Print a help text and exit.
.IP \[bu] 2
\f[B]\-\-version\f[]
.PD 0
.P
.PD
Print an info text containing the compilation date and the program
version.
.IP \[bu] 2
\f[B]\-o|\-\-colors\f[]
.PD 0
.P
.PD
Enable coloured output where available.
Will mess up pipelines, so be careful.
.IP \[bu] 2
\f[B]\-q|\-\-quiet\f[]
.PD 0
.P
.PD
Suppress all output to \f[B]STDOUT\f[] and \f[B]STDERR\f[].
However, if the program aborted due to system error (such as a failed
memory allocation), it will still error out with a debug message.
In theory, this should never happen of course.
.SS OPTIONS INVOLVING TAGS
.IP \[bu] 2
\f[B]\-v|\-\-verify [\f[I]\-\-\f[]] \f[I]FILE\f[] \f[]
.PD 0
.P
.PD
Verify the integrity of \f[I]FILE\f[] using the tag in its filename.
.IP \[bu] 2
\f[B]\-t|\-\-tag [\-s|\-\-strip\-tag] [\-D|\-\-dot INT] [\f[I]\-\-\f[]]
\f[I]FILE\f[] \f[]
.PD 0
.P
.PD
Tag \f[I]FILE\f[] with the hexstring of its CRC32 checksum.
Fails if \f[I]FILE\f[] has already been tagged.
If \f[I]\-s\f[] is present, strip any existing tag before appending the
new one.
By default, the tag is inserted into the filename in front of the last
dot.
In the case of multi\-part suffixes like .tar.gz, it is desireable to
insert the tag in another position.
Using the \-\-dot option, the tag will be inserted in front of the INTth
dot in the filename.
Negative indices are also possible and indicate the INTth\-last dot to
be used.
Zero \[aq]0\[aq] is a nonsense argument to the option and will be
ignored.
\-\-dot defaults to \-1.
.IP \[bu] 2
\f[B]\-r|\-\-remove\-tag [\f[I]\-\-\f[]] \f[I]FILE\f[] \f[]
.PD 0
.P
.PD
If \f[I]FILE\f[] is tagged, remove the tag.
.IP \[bu] 2
\f[B]\-X|\-\-verify\-hex \f[I]TAG\f[] [\-n|\-\-numerical]
[\f[I]\-\-\f[]] \f[I]FILE\f[] \f[]
.PD 0
.P
.PD
Given a CRC32 code \f[I]TAG\f[], check whether \f[I]FILE\f[] has the
CRC32 code \f[I]TAG\f[].
\f[I]TAG\f[] is expected to be a hexadecimal number, unless
\f[B]\-\-numerical\f[] is active, in which case it is expected to be a
decimal number.
.SS OPTIONS INVOLVING DATABASES
.IP \[bu] 2
\f[B]\-p|\-\-print [\-R|\-\-realpath] \f[I]DB\-FILE\f[] \f[]
.PD 0
.P
.PD
Print all records in \f[I]DB\-FILE\f[] to stdout and exit.
If \-\-realpath is specified, the paths will be printed as absolute
paths.
.IP \[bu] 2
\f[B]\-C|\-\-create\-db \f[I]DB\-FILE\f[] [\-a|\-\-append]
[\-R|\-\-realpath] [\f[I]\-\-\f[]] \f[I]FILE\-LIST\f[] \f[]
.PD 0
.P
.PD
For a every \f[I]FILE\f[] in \f[I]FILE\-LIST\f[], record its path and
CRC32 checksum in the database \f[I]DB\-FILE\f[].
Relative filepaths are used.
If the same path appears more than once in \f[I]FILE\-LIST\f[], the
record will be overwritten.
Any pre\-existing \f[I]DB\-FILE\f[] will be overwritten unless
\f[I]\-a\f[] (append) is present.
If the \f[I]\-R\f[] (realpath) option was specified, the absolute file
paths will be used instead the ones on the command line when storing the
file entries.
.IP \[bu] 2
\f[B]\-V|\-\-verify\-db \f[I]DB\-FILE\f[] [\-x|\-\-prefer\-hexstring]
[\f[I]\-\-\f[]] [\f[I]FILE\-LIST\f[]] \f[]
.PD 0
.P
.PD
For every \f[I]FILE\f[] in \f[I]FILE\-LIST\f[], check if it has a record
in \f[I]DB\-FILE\f[] and if yes, verify if its current CRC32 checksum
matches the checksum recorded in the database.
If \f[I]\-x\f[] is present, for any tagged \f[I]FILE\f[], the
hexstring\-encoded CRC32 checksum is used for comparison instead of
computing a reference sum.
.PD 0
.P
.PD
If \f[I]FILE\-LIST\f[] is absent, for every \f[I]FILE\f[] recorded in
\f[I]DB\-FILE\f[], check whether the file still exists and if yes,
verify its current CRC32 checksum against the one recorded in the
database.
.IP \[bu] 2
\f[B]\-m|\-\-merge \f[I]DB\-FILE\f[] [\-a|\-\-append] [\-R|\-\-realpath]
[\f[I]\-\-\f[]] \f[I]DB\-FILE\-LIST\f[] \f[]
.PD 0
.P
.PD
For every \f[I]DB\-FILE\f[] in \f[I]DB\-FILE\-LIST\f[], merge its the
contents into the database supplied as an paramter to the \-m option.
Duplicate records will overwrite each other in the order they appear.
If \-\-realpath is specified, all paths in the resulting database will
be absolute paths.
For this to work, relative paths in the respective databases must still
be valid relative to the current working directory.
.IP \[bu] 2
\f[B]\-d|\-\-delete \f[I]DB\-FILE\f[] [\f[I]\-\-\f[]]
\f[I]FILE\-LIST\f[] \f[]
.PD 0
.P
.PD
Delete every \f[I]FILE\f[] in \f[I]FILE\-LIST\f[] from \f[I]DB\-FILE\f[]
if it exists.
.IP \[bu] 2
\f[B]\-P|\-\-to\-realpath [\f[I]\-\-\f[]] \f[I]DB\-FILE\-LIST\f[] \f[]
.PD 0
.P
.PD
Convert all paths in the given databases into absolute paths if
possible.
For this to work, any relative paths must still be valid.
.SH EXAMPLES
.PP
No manual be complete without examples.
.PP
Given a file \f[I]archive.tar\f[], calculate its CRC32:
.RS
.PP
crctk \-c \f[I]archive.tar\f[]
.RE
.PP
Affix a CRC32 tag to the filename, renaming the file to
\f[I]archive[AABBCC99].tar\f[]:
.RS
.PP
crctk \-t \f[I]archive.tar\f[]
.RE
.PP
Check whether \f[I]archive[AABBCC99].tar\f[] matches the CRC32 sum
\f[B]AACCBB99\f[] (it doesn\[aq]t):
.RS
.PP
crctk \-X 0xaaccbb99 \f[I]archive[AABBCC99].tar\f[]
.RE
.PP
Verify whether the CRC32 code in the filename matches the file:
.RS
.PP
crctk \-v \f[I]archive[AABBCC99].tar\f[]
.RE
.PP
Record the paths and CRC32 codes of two files in a database:
.RS
.PP
crctk \-C db.crctk \f[I]file1.bin\f[] \f[I]file2.bin\f[]
.RE
.PP
Verify whether all files in the database still match the recorded CRC32
code:
.RS
.PP
crctk \-V db.crctk
.RE
.PP
Check whether a file is recorded in the database and if yes, whether its
CRC32 code matches the recorded one:
.RS
.PP
crctk \-V db.crctk \f[I]file2.bin\f[]
.RE
.PP
Create a database but if the files do already have CRC32 codes recorded
in their filenames, use these in the database record instead of
calculating the CRC32 codes again.
Don\[aq]t truncate but append to an existing database:
.RS
.PP
crctk \-axC db.crctk \f[I]file3[aaccbbee].bin\f[]
.RE
.SH AUTHORS
Jens Oliver John <dev@2ion.de>.


ChangeLog
2014-04-01  gettextize  <bug-gnu-gettext@gnu.org>

	* Makefile.am (SUBDIRS): Add po.
	(ACLOCAL_AMFLAGS): New variable.
	(EXTRA_DIST): New variable.
	* configure.ac (AC_CONFIG_FILES): Add po/Makefile.in.


src/crctk.c
/*
 * crctk - CRC32 Hexstring Toolkit
 * Copyright (C) 2014 Jens Oliver John <dev@2ion.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * */

#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>

#include "command_calc.h"
#include "command_calc_batch.h"
#include "command_check.h"
#include "command_check_batch.h"
#include "command_check_hexstring.h"
#include "command_delete.h"
#include "command_help.h"
#include "command_idle.h"
#include "command_list.h"
#include "command_merge.h"
#include "command_remove_tag.h"
#include "command_tag.h"
#include "command_to_realpaths.h"
#include "crctk.h"
#include "util.h"

inline static void free_if_flagged(void *p, int *flag) {
  if(*flag == 1)
    free(p);
}

/* from crctk.h */

const char *crcregex = "[[:xdigit:]]\\{8\\}";
const char *crcregex_stripper =
  "[[:punct:]]\\?[[:xdigit:]]\\{8\\}[[:punct:]]\\?";
const char *dbiofile = NULL;
const char *hexarg = "00000000";
int dotidx = -1;
int flag_use_colors = 0;
int flag_be_quiet = 0;

static const char *optstring_short = "+X:xtnvV:hSsRrC:cD:d:e:Ppam:qo";
static const struct option options_long[] = {
  { "verify", no_argument, NULL, 'v' },
  { "verify-db", required_argument, NULL, 'V' },
  { "hexstring", no_argument, NULL, 'x' },
  { "calc", no_argument, NULL, 'c' },
  { "numerical", no_argument, NULL, 'n' },
  { "create-db", required_argument, NULL, 'C' },
  { "delete", required_argument, NULL, 'd' },
  { "append", no_argument, NULL, 'a' },
  { "print", no_argument, NULL, 'p' },
  { "tag", no_argument, NULL, 't' },
  { "dot", required_argument, NULL, 'D' },
  { "strip-tag", no_argument, NULL, 's' },
  { "remove-tag", no_argument, NULL, 'r' },
  { "tag-regex", required_argument, NULL, 'e' },
  { "help", no_argument, NULL, 'h' },
  { "version", no_argument, NULL, 'J' },
  { "merge", required_argument, NULL, 'm' },
  { "realpath", no_argument, NULL, 'R' },
  { "to-realpath", no_argument, NULL, 'P' },
  { "colors", no_argument, NULL, 'o' },
  { "quiet", no_argument, NULL, 'q' },
  { "verify-hex", required_argument, NULL, 'X' },
  { 0, 0, 0, 0 }
};

int main(int argc, char **argv) {
  int opt = 0;
  int cmdflags = 0;
  int do_free_dbiofile = 0;
  int do_free_hexarg = 0;
  int do_free_crcregexstripper = 0;
  int ret = EXIT_SUCCESS;
  CommandFunction cmd = command_idle;

#define ASSIGN_OPTARG_IF_NULL(cptr, fptr) if((cptr)==NULL){\
  (cptr) = strdup(optarg);\
  (*fptr) = 1;\
}
  while((opt = getopt_long(argc, argv, optstring_short,
         options_long, NULL)) != -1)
    switch(opt) {
      case 's': cmdflags |= TAG_ALLOW_STRIP;
                break;
      case 'x': cmdflags |= CHECK_BATCH_PREFER_HEXSTRING;
                break;
      case 'a': cmdflags |= APPEND_TO_DB;
                break;
      case 'n': cmdflags |= CALC_PRINT_NUMERICAL;
                break;
      case 'R': cmdflags |= USE_REALPATH;
                break;
      case 'p': cmd = command_list_db;
                break;
      case 'V': ASSIGN_OPTARG_IF_NULL(dbiofile, &do_free_dbiofile);
                cmd = command_check_batch;
                break;
      case 'e': crcregex_stripper = strdup(optarg);
                do_free_crcregexstripper = 1;
                break;
      case 'c': cmd = command_calc;
                break;
      case 'C': ASSIGN_OPTARG_IF_NULL(dbiofile, &do_free_dbiofile);
                cmd = command_calc_batch;
                break;
      case 'P': cmd = command_to_realpaths;
                break;
      case 'r': cmd = command_remove_tag;
                break;
      case 'v': cmd = command_check;
                break;
      case 'X': hexarg = strdup(optarg);
                do_free_hexarg = 1;
                cmd = command_check_hexstring;
                break;
      case 't': cmd = command_tag;
                break;
      case 'h': cmd = command_help;
                break;
      case 'm': dbiofile = strdup(optarg);
                do_free_dbiofile = 1;
                cmd = command_merge;
                break;
      case 'J': puts("crctk version: " PACKAGE_VERSION);
                puts("Compiled on: " __DATE__ " " __TIME__);
                return EXIT_SUCCESS;
                break;
      case 'd': dbiofile = strdup(optarg);
                do_free_dbiofile = 1;
                cmd = command_delete;
                break;
      case 'D': dotidx = strtol((const char*)optarg, NULL, 0xA);
                if(dotidx == LONG_MIN && errno == ERANGE)
                    LERROR(EXIT_FAILURE, errno, "Integer argument to the -D option underflows");
                if(dotidx == LONG_MAX && errno == ERANGE)
                    LERROR(EXIT_FAILURE, errno, "Integer argument to the -D option overflows");
                if(dotidx == 0) {
                  LOG_INFO("Argument to --dot, -D is zero which makes no sense: defaulting to -1 instead");
                  dotidx = -1;
                }
                break;
      case 'q': flag_be_quiet = 1;
                break;
      case 'o': flag_use_colors = 1;
                break;
      case 0:   continue;
                break;
      default:  return EXIT_FAILURE;
    } // switch
#undef ASSIGN_OPTARG_IF_NULL
  if(optind >= argc &&
      cmd != command_check_batch &&
      cmd != command_list_db && 
      cmd != command_help) {
    log_failure("crctk", "Too few arguments. Consult --help "
          "for usage information.");
    return EXIT_FAILURE;
  }

  /* pick up environment variables */
  if(getenv("CRCTK_ENABLE_COLORS"))
    flag_use_colors = 1;

  ret = cmd(argc, argv, optind, cmdflags);

  free_if_flagged((void*)dbiofile, &do_free_dbiofile);
  free_if_flagged((void*)hexarg, &do_free_hexarg);
  free_if_flagged((void*)crcregex_stripper, &do_free_crcregexstripper);

  return ret;
}
